File: 240522/네 방향 탈출 가능 여부 판별하기/determine-escapableness-with-4-ways.js
const fs = require('fs')
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n')

const [n, m] = input[0].split(" ").map(x => parseInt(x))
const arr = []
for (i = 1; i < input.length; i++) {
    x = input[i].split(" ").map(x => parseInt(x))
    arr.push(x)
}

let visited = Array.from({length : n}, () => Array(m).fill(0))
const dir = [[0, 1], [1, 0], [0, -1], [-1, 0]]

function bfs() {

    q = [[0, 0]]

    while (q.length) {
        const [x, y] = q.shift()
        
        dir.forEach((d) => {
            xd = x + d[0]
            yd = y + d[1]

            if (0 <= xd && xd < n && 0 <= yd && yd < m) {
                if (arr[xd][yd] && visited[xd][yd] == 0) {
                    visited[xd][yd] = 1
                    q.push([xd, yd])
                }
            }
        })
    }
}

bfs()
if (visited[n-1][m-1]) {
    console.log(1)
} else {
    console.log(0)
}

File: 240521/네 방향 탈출 가능 여부 판별하기/determine-escapableness-with-4-ways.py
n, m = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(n)]
visited = [[0 for _ in range(m)] for _ in range(n)]
dir = [(0, 1), (1, 0), (0, -1), (-1, 0)]
def bfs(sn, sm):

    q = [(sn, sm)]
    while q:
        x, y = q.pop(0)
        for dn, dm in dir:
            xn, ym  = x + dn, y + dm 
            if 0 <= xn < n and 0 <= ym < m:
                if arr[xn][ym] == 1 and visited[xn][ym] == 0:
                    visited[xn][ym] = 1
                    q.append((xn, ym))

bfs(0, 0)
if visited[n-1][m-1]:
    print(1)
else:
    print(0)